
Colony PCR and Sanger Sequencing
=================================

## Authorship

Janani Hariharan and Roli Wilhelm (2018)

## Description
This is a general colony PCR protocol to generate amplicons that can be sent for Sanger sequencing. Specific PCR conditions will vary for individual sets of PCR primers, but the pre- and post-PCR steps can be used with any PCR protocol. 

## Method

1.	Prepare PCR mix as described in the [PCR protocol](https://github.com/buckleylab/Buckley_lab_protocols/blob/master/PCR/PCR_with_barcoded_primers.md).
2.	Scrape a colony onto a toothpick and transfer to the well/tube containing the PCR mix. Make sure to pick out an individual colony. 
3.	If using a 96-well skirted plate, seal the plate with Microseal B and flatten with roller to ensure adequate sealing. 
4.	Insert plate into thermocycler and close lid. Set the PCR conditions as appropriate to primer set.

## Size Selection

Verify the size of your amplicon by running it on an agarose gel prepared with 0.5x TE buffer as described in steps 2-5 [here](https://github.com/buckleylab/Buckley_lab_protocols/blob/master/gel_electrophoresis/gel_extraction.md). For PCR products, 1 uL + 5 uL of loading dye can be run on a 1% gel (with SYBR Safe). 

## DNA Extraction 
 
DNA can be extracted from the gel slice using the Wizard SV Gel and PCR Clean-Up Kit as described in the gel electrophoresis [protocol](https://github.com/buckleylab/Buckley_lab_protocols/blob/master/gel_electrophoresis/gel_extraction.md). Alternatively, follow the below protocol for AmPureXP MagBead-based clean-up of the PCR product.

## AmPure XP Magnetic Bead Cleanup

1. Prepare fresh 20% Polyethyleneglycol (PEG) 8000 with 0.9M NaCl.

2. Prepare fresh 75% ethanol.

3. Mix 20ul of beads with 70 uL of PEG and, lastly, add 100ul DNA (make this up by combining volume of your PCR product with DNase-free water) in 96-well plate (mix gently by pipetting up and down).
	Note: Make sure the 96-well plate is of appropriate fit for magnetic plate.

4. Incubate mixture for 5 minutes.

5. Place 96-well plate on magnetic plate and allow beads to settle for 5 minutes.

6. Discard supernatant.

7. Wash step: Add 200ul 75% ethanol, allow beads to settle, remove ethanol and repeat once.

8. Allow beads to dry for ~2-3 minutes (until all ethanol evaporates).

9. As soon as you notice any cracking in beads, remove 96-well plate from magnetic plate and resuspend all samples in 16 ul of DNase-free water.

10. Place 96-well plate on magnetic plate and allow beads to settle. Pipette out water (which should now contain DNA) from each well into sterile, labeled tubes. 

If desired, DNA can be quantified prior to sequencing using the PicoGreen protocol described [here](https://github.com/buckleylab/Buckley_Lab_SIP_project_protocols/blob/master/nucleotide_conc/picogreen.md). This is recommended if the PCR product is being sequenced for the first time. 

## Sample Preparation for Sanger Sequencing

1.	Dilute primer to 25 pmol/uL (from 10 uM working stock) with DNase-free water (only one primer may be added per sample for Sanger sequencing).

2.	Add 2 uL of this solution to 16 uL of cleaned PCR product in specially marked sequencing tubes. For additional info on the volumes of DNA and primer required for sequencing, refer to the [BRC Handbook](http://www.biotech.cornell.edu/node/556). 


